### PR TITLE
Feature/494 use file resource

### LINKED ISSUE #494: "Create CSV file" to use a Resource instead of a java.util.File
Instead of restricting ourselves to writing CSV _files_ we should rather change the "Create CSV file" analyzer to use a org.apache.metamodel.util.Resource instead of a java.util.File. That way we will enable also writing CSV content in other file systems than just local disk.

### PR DESCRIPTION
**Pull request by ClaudiaPHI**
Fixes #494 

There are still problems with writing to a HDSF. I get the following stacktrace from a hadoop cluster  formed by a single node. 

I tried to configure hadoop not to replicate the file but I did not succeed to get rid of the <b>DataStreamer Exception </b> 

```
WARN  16:23:51 DFSClient - DataStreamer Exception
java.io.IOException: Failed to replace a bad datanode on the existing pipeline due to no more good datanodes being available to try. (Nodes: current=[127.0.0.1:50010], original=[127.0.0.1:50010]). The current failed datanode replacement policy is DEFAULT, and a client may configure this via 'dfs.client.block.write.replace-datanode-on-failure.policy' in its configuration.
	at org.apache.hadoop.hdfs.DFSOutputStream$DataStreamer.findNewDatanode(DFSOutputStream.java:1040)
	at org.apache.hadoop.hdfs.DFSOutputStream$DataStreamer.addDatanode2ExistingPipeline(DFSOutputStream.java:1106)
	at org.apache.hadoop.hdfs.DFSOutputStream$DataStreamer.setupPipelineForAppendOrRecovery(DFSOutputStream.java:1253)
	at org.apache.hadoop.hdfs.DFSOutputStream$DataStreamer.run(DFSOutputStream.java:594)
ERROR 16:23:51 DFSClient - Failed to close inode 16457
java.io.IOException: Failed to replace a bad datanode on the existing pipeline due to no more good datanodes being available to try. (Nodes: current=[127.0.0.1:50010], original=[127.0.0.1:50010]). The current failed datanode replacement policy is DEFAULT, and a client may configure this via 'dfs.client.block.write.replace-datanode-on-failure.policy' in its configuration.
	at org.apache.hadoop.hdfs.DFSOutputStream$DataStreamer.findNewDatanode(DFSOutputStream.java:1040)
	at org.apache.hadoop.hdfs.DFSOutputStream$DataStreamer.addDatanode2ExistingPipeline(DFSOutputStream.java:1106)
	at org.apache.hadoop.hdfs.DFSOutputStream$DataStreamer.setupPipelineForAppendOrRecovery(DFSOutputStream.java:1253)
	at org.apache.hadoop.hdfs.DFSOutputStream$DataStreamer.run(DFSOutputStream.java:594)
WARN  16:23:52 ErrorAwareAnalysisListener - errorUnknown(ImmutableAnalysisJob[sourceColumns=8,filterJobs=0,transformerJobs=0,analyzerJobs=1],org.apache.metamodel.MetaModelException: org.apache.hadoop.ipc.RemoteException(org.apache.hadoop.hdfs.protocol.AlreadyBeingCreatedException): Failed to APPEND_FILE /user/vagrant/test.csv for DFSClient_NONMAPREDUCE_-296689498_137 on 192.168.33.1 because this file lease is currently owned by DFSClient_NONMAPREDUCE_1140524127_137 on 192.168.33.1
	at org.apache.hadoop.hdfs.server.namenode.FSNamesystem.recoverLeaseInternal(FSNamesystem.java:2933)
	at org.apache.hadoop.hdfs.server.namenode.FSNamesystem.appendFileInternal(FSNamesystem.java:2685)
	at org.apache.hadoop.hdfs.server.namenode.FSNamesystem.appendFileInt(FSNamesystem.java:2985)
	at org.apache.hadoop.hdfs.server.namenode.FSNamesystem.appendFile(FSNamesystem.java:2952)
	at org.apache.hadoop.hdfs.server.namenode.NameNodeRpcServer.append(NameNodeRpcServer.java:653)
	at org.apache.hadoop.hdfs.protocolPB.ClientNamenodeProtocolServerSideTranslatorPB.append(ClientNamenodeProtocolServerSideTranslatorPB.java:421)
	at org.apache.hadoop.hdfs.protocol.proto.ClientNamenodeProtocolProtos$ClientNamenodeProtocol$2.callBlockingMethod(ClientNamenodeProtocolProtos.java)
	at org.apache.hadoop.ipc.ProtobufRpcEngine$Server$ProtoBufRpcInvoker.call(ProtobufRpcEngine.java:616)
	at org.apache.hadoop.ipc.RPC$Server.call(RPC.java:969)
	at org.apache.hadoop.ipc.Server$Handler$1.run(Server.java:2049)
	at org.apache.hadoop.ipc.Server$Handler$1.run(Server.java:2045)
	at java.security.AccessController.doPrivileged(Native Method)
	at javax.security.auth.Subject.doAs(Subject.java:415)
	at org.apache.hadoop.security.UserGroupInformation.doAs(UserGroupInformation.java:1657)
	at org.apache.hadoop.ipc.Server$Handler.run(Server.java:2043)
) 
WARN  16:23:52 ErrorAwareAnalysisListener - Exception stack trace:
org.apache.metamodel.MetaModelException: org.apache.hadoop.ipc.RemoteException(org.apache.hadoop.hdfs.protocol.AlreadyBeingCreatedException): Failed to APPEND_FILE /user/vagrant/test.csv for DFSClient_NONMAPREDUCE_-296689498_137 on 192.168.33.1 because this file lease is currently owned by DFSClient_NONMAPREDUCE_1140524127_137 on 192.168.33.1
	at org.apache.hadoop.hdfs.server.namenode.FSNamesystem.recoverLeaseInternal(FSNamesystem.java:2933)
	at org.apache.hadoop.hdfs.server.namenode.FSNamesystem.appendFileInternal(FSNamesystem.java:2685)
	at org.apache.hadoop.hdfs.server.namenode.FSNamesystem.appendFileInt(FSNamesystem.java:2985)
	at org.apache.hadoop.hdfs.server.namenode.FSNamesystem.appendFile(FSNamesystem.java:2952)
	at org.apache.hadoop.hdfs.server.namenode.NameNodeRpcServer.append(NameNodeRpcServer.java:653)
	at org.apache.hadoop.hdfs.protocolPB.ClientNamenodeProtocolServerSideTranslatorPB.append(ClientNamenodeProtocolServerSideTranslatorPB.java:421)
	at org.apache.hadoop.hdfs.protocol.proto.ClientNamenodeProtocolProtos$ClientNamenodeProtocol$2.callBlockingMethod(ClientNamenodeProtocolProtos.java)
	at org.apache.hadoop.ipc.ProtobufRpcEngine$Server$ProtoBufRpcInvoker.call(ProtobufRpcEngine.java:616)
	at org.apache.hadoop.ipc.RPC$Server.call(RPC.java:969)
	at org.apache.hadoop.ipc.Server$Handler$1.run(Server.java:2049)
	at org.apache.hadoop.ipc.Server$Handler$1.run(Server.java:2045)
	at java.security.AccessController.doPrivileged(Native Method)
	at javax.security.auth.Subject.doAs(Subject.java:415)
	at org.apache.hadoop.security.UserGroupInformation.doAs(UserGroupInformation.java:1657)
	at org.apache.hadoop.ipc.Server$Handler.run(Server.java:2043)

	at org.apache.metamodel.util.HdfsResource.wrapException(HdfsResource.java:264)
	at org.apache.metamodel.util.HdfsResource.append(HdfsResource.java:157)
	at org.apache.metamodel.csv.CsvUpdateCallback.writeRow(CsvUpdateCallback.java:112)
	at org.apache.metamodel.csv.CsvInsertBuilder.execute(CsvInsertBuilder.java:37)
	at org.datacleaner.output.AbstractMetaModelOutputWriter$1.run(AbstractMetaModelOutputWriter.java:95)
	at org.apache.metamodel.csv.CsvDataContext.executeUpdate(CsvDataContext.java:429)
	at org.datacleaner.output.AbstractMetaModelOutputWriter.flushBuffer(AbstractMetaModelOutputWriter.java:86)
	at org.datacleaner.output.AbstractMetaModelOutputWriter.close(AbstractMetaModelOutputWriter.java:106)
	at org.datacleaner.extension.output.AbstractOutputWriterAnalyzer.getResult(AbstractOutputWriterAnalyzer.java:58)
	at org.datacleaner.extension.output.AbstractOutputWriterAnalyzer.getResult(AbstractOutputWriterAnalyzer.java:1)
	at org.datacleaner.job.tasks.CollectResultsTask.execute(CollectResultsTask.java:60)
	at org.datacleaner.job.concurrent.TaskRunnable.run(TaskRunnable.java:61)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: org.apache.hadoop.ipc.RemoteException(org.apache.hadoop.hdfs.protocol.AlreadyBeingCreatedException): Failed to APPEND_FILE /user/vagrant/test.csv for DFSClient_NONMAPREDUCE_-296689498_137 on 192.168.33.1 because this file lease is currently owned by DFSClient_NONMAPREDUCE_1140524127_137 on 192.168.33.1
	at org.apache.hadoop.hdfs.server.namenode.FSNamesystem.recoverLeaseInternal(FSNamesystem.java:2933)
	at org.apache.hadoop.hdfs.server.namenode.FSNamesystem.appendFileInternal(FSNamesystem.java:2685)
	at org.apache.hadoop.hdfs.server.namenode.FSNamesystem.appendFileInt(FSNamesystem.java:2985)
	at org.apache.hadoop.hdfs.server.namenode.FSNamesystem.appendFile(FSNamesystem.java:2952)
	at org.apache.hadoop.hdfs.server.namenode.NameNodeRpcServer.append(NameNodeRpcServer.java:653)
	at org.apache.hadoop.hdfs.protocolPB.ClientNamenodeProtocolServerSideTranslatorPB.append(ClientNamenodeProtocolServerSideTranslatorPB.java:421)
	at org.apache.hadoop.hdfs.protocol.proto.ClientNamenodeProtocolProtos$ClientNamenodeProtocol$2.callBlockingMethod(ClientNamenodeProtocolProtos.java)
	at org.apache.hadoop.ipc.ProtobufRpcEngine$Server$ProtoBufRpcInvoker.call(ProtobufRpcEngine.java:616)
	at org.apache.hadoop.ipc.RPC$Server.call(RPC.java:969)
	at org.apache.hadoop.ipc.Server$Handler$1.run(Server.java:2049)
	at org.apache.hadoop.ipc.Server$Handler$1.run(Server.java:2045)
	at java.security.AccessController.doPrivileged(Native Method)
	at javax.security.auth.Subject.doAs(Subject.java:415)
	at org.apache.hadoop.security.UserGroupInformation.doAs(UserGroupInformation.java:1657)
	at org.apache.hadoop.ipc.Server$Handler.run(Server.java:2043)

	at org.apache.hadoop.ipc.Client.call(Client.java:1468)
	at org.apache.hadoop.ipc.Client.call(Client.java:1399)
	at org.apache.hadoop.ipc.ProtobufRpcEngine$Invoker.invoke(ProtobufRpcEngine.java:232)
	at com.sun.proxy.$Proxy72.append(Unknown Source)
	at org.apache.hadoop.hdfs.protocolPB.ClientNamenodeProtocolTranslatorPB.append(ClientNamenodeProtocolTranslatorPB.java:313)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.apache.hadoop.io.retry.RetryInvocationHandler.invokeMethod(RetryInvocationHandler.java:187)
	at org.apache.hadoop.io.retry.RetryInvocationHandler.invoke(RetryInvocationHandler.java:102)
	at com.sun.proxy.$Proxy73.append(Unknown Source)
	at org.apache.hadoop.hdfs.DFSClient.callAppend(DFSClient.java:1767)
	at org.apache.hadoop.hdfs.DFSClient.append(DFSClient.java:1803)
	at org.apache.hadoop.hdfs.DFSClient.append(DFSClient.java:1796)
	at org.apache.hadoop.hdfs.DistributedFileSystem$4.doCall(DistributedFileSystem.java:323)
	at org.apache.hadoop.hdfs.DistributedFileSystem$4.doCall(DistributedFileSystem.java:319)
	at org.apache.hadoop.fs.FileSystemLinkResolver.resolve(FileSystemLinkResolver.java:81)
	at org.apache.hadoop.hdfs.DistributedFileSystem.append(DistributedFileSystem.java:319)
	at org.apache.hadoop.fs.FileSystem.append(FileSystem.java:1163)
	at org.apache.metamodel.util.HdfsResource.append(HdfsResource.java:150)
	... 13 more

```

### CODE REVIEW

**Comment by LosD** [comment]
Sooo, I guess this shouldn't be merged yet?

**Comment by ClaudiaPHI** [comment]
No. Until we do not figure out how to solve the exception

File: engine/core/src/main/java/org/datacleaner/util/sort/SortMergeWriter.java [hunk 1]
```diff
@@ -174,6 +177,28 @@ public File write(String filename) {
...
     /**
+     * @param resource
+     * @return the written count of rows
+     */
+    public int write(final Resource file) {
+        W writer = null;
+        ObjectInputStream[] tempFileObjectInputStreams = null;
+        try {
+            writer = createWriter(file);
+            return processRows(writer, tempFileObjectInputStreams);
```
Comment (kaspersorensen) [inline engine/core/src/main/java/org/datacleaner/util/sort/SortMergeWriter.java]: I don't get this. The way I see it 'tempFileObjectInputStreams' will always be null here, and also in the write(File) method below.

I would imagine that when we add a write(Resource) method, then we can implement the write(File) method simply by calling write(new FileResource(file)) and then reuse the complete write method code that way.

**Comment by kaspersorensen** [comment]
All those createWriter method calls that return null ... It smells of something fundamentally wrong. Probably a better approach would be to re-implement the SortMergeWriter based on Resource instead of File. You wouldn't need to retain backwards compatibility IMO, but for the most important method you could even do it simply by wrapping the File argument in a FileResource.

**Comment by ClaudiaPHI** [comment]
The branch fails because it is built with a metamodel snapshot. Please check the issue on jira.

**Comment by kaspersorensen** [comment]
Superceeded by #588

**Pull request closed without merging**